### DISCUSSION TITLE
[Messenger] Make the `sync://` transport safe

### DISCUSSION THREAD

**Discussion by bpolaszek** [symfony/symfony#49869] [Polls]
Hello there,

A few weeks ago, one of the projects I was working on had issues with concurrent operations on a `messenger` queue (context: several workers on several machines, using the `redis` transport). As _some_ messages had to be processed sequentially (e.g. not simultaneously by several workers), we decided to temporarily switch the `redis` transport to `sync://`.

Surprisingly, with `sync://`:
- When an exception is raised during the processing of the message, **the whole process crashes**
- As a consequence, messages are lost and **can't be retried**

So a synchronous transport cannot be a drop-in replacement of an asynchronous one,  even though they share the same contract (that's why I had the same expectations). As a workaround, I crafted a [`safe-sync://`](https://github.com/bpolaszek/safe-sync-transport) transport*: it acts synchronously, but behaves like an async transport would: whenever a message processing fails, it retries it, then eventually moves it to a `failed` transport, and the PHP process can come to its expected end.

_* It's a proof-of-concept that I never used in production - we found another, domain-related solution_

**Comment by ghost** [comment]
That's an interesting problem. AFAIK sync transport was created mainly for testing purpose. Basically sync transport is just like no transport at all, it's a synchronous code execution so at the first glance it should not be surprising that _the whole process crashes_. However I agree that we can discuss whether the behavior of sync transport in context of failed messages should be redefined.

**Comment by thiagorb** [comment]
I llike the option to make it configurable, but I would prefer if the default behavior were changed. Another problem with the current behavior is that the event WorkerMessageFailedEvent is never triggered for the sync transport.

**Reply by bpolaszek** [reply]
That's exactly [the point](https://github.com/bpolaszek/safe-sync-transport/blob/main/src/Transport/SafeSyncTransport.php#L92)!

**Reply by thiagorb** [reply]
@bpolaszek I can't read this source. Is your repo private?

**Reply by bpolaszek** [reply]
Oh, silly me! 😅 

Indeed, I didn't notice I forgot to change the visibility.
That's fixed!

**Comment by alexander-schranz** [comment]
Just stumble over this. In my case the `MessageBus` is in sync case used as a `CommandBus` and so it is expected that a thrown exception does throw the Exception out and somebody need to catch it and not behave like a async bus. Even in my cases the `HandlerFailedException` is [unpacked in a custom midddleware and so the real exception](
https://github.com/sulu/messenger/blob/main/src/Infrastructure/Symfony/Messenger/UnpackExceptionMiddleware/UnpackExceptionMiddleware.php) is thrown outside. Mostly ModelNotFoundException which are catched in the controller and response 404.

**Reply by MatTheCat** [reply]
I guess #52300 would make sense to you then? :grin:

**Comment by Sam152** [comment]
For my use case, I think it makes sense to retain the current behaviour of crashing the whole process or making it configurable, so this behaviour can be retained. 

I have message handlers where a synchronous handler records some domain interaction, and a series of queued asynchronous handlers provision a number of resources that follow from that interaction. Occasionally the synchronous handler may throw some invariant exception, where all of the resources provisioned asynchronously should be halted and discarded.

A "safe-sync://" transport would be a good way to support both use cases.

**Comment by ro0NL** [comment]
sync:// dsn + retry_strategy/failure_transport should work as expected IMHO

**Comment by linaori** [comment]
My use-case is to make an automatic retry mechanism for deadlocks and lock wait timeouts using the `sync://` in a command bus. With this I also run into the issue that throwing an exception does not go into the retry as configured. Not using a transport currently also doesn't let you use a retry unless you make a custom middleware to deal with this.
